static/js/smart-plug.js

```diff
@@ -0,0 +1,184 @@
+};
+
+/**
+ * HTML view for smart plug.
+ */
+SmartPlug.prototype.htmlView = function() {
+  return `<a href="${this.href}">
+    <div class="thing smart-plug">
+      ${this.iconView()}
+      <span class="thing-name">${this.name}</span>
+    </div>
+  </a>`;
+};
+
+/**
+ * HTML detail view for smart plug.
+ */
+SmartPlug.prototype.htmlDetailView = function() {
+  let detailsHTML = '';
+  for (let prop in this.details) {
+    detailsHTML += this.details[prop].view();
+  }
+
+  return `<div class="smart-plug-container">
+    <div class="thing">
+      ${this.iconView()}
+    </div>
+    ${detailsHTML}
+  </div>`;
+};
+
+/**
+ * SVG view for smart plug.
+ */
+SmartPlug.prototype.svgView = function() {
+  return '<g transform="translate(' + this.x + ',' + this.y + ')"' +
+         '  dragx="' + this.x + '" dragy="' + this.y + '"' +
+         '  class="floorplan-thing">' +
+         '  <a href="' + this.href +'" class="svg-thing-link">' +
+         '    <circle cx="0" cy="0" r="5" class="svg-thing-icon" />' +
+         '    <image x="-2.5" y="-2.5" width="5" height="5" ' +
+         '      xlink:href="/images/smart-plug-off.svg" />' +
+         '    <text x="0" y="8" text-anchor="middle" class="svg-thing-text">' +
+                this.name.substring(0, 7) +
+         '    </text>' +
+         '  </a>' +
+         '</g>';
+};
+
+/**
+ * Update the status of the smart plug.
+ */
+SmartPlug.prototype.updateStatus = function() {
+  var opts = {
+    headers: {
+      'Authorization': `Bearer ${window.API.jwt}`,
+      'Accept': 'application/json'
+    }
+  };
+
+  fetch(this.onPropertyUrl, opts).then(response => {
+    return response.json();
+  }).then(response => {
+    this.onPropertyStatus(response);
+    return fetch(this.powerPropertyUrl, opts);
+  }).then(response => {
+    return response.json();
+  }).then(response => {
+    this.onPropertyStatus(response);
+  }).catch(error => {
+    console.error('Error fetching smart plug status ' + error);
+  });
+};
+
+/**
+ * Handle a 'propertyStatus' message.
+ * @param {Object} properties - property data
+ */
+SmartPlug.prototype.onPropertyStatus = function(data) {
+  if (data.hasOwnProperty('on')) {
+    this.updateOn(data.on);
+  }
+  if (data.hasOwnProperty('instantaneousPower')) {
+    this.updatePower(data.instantaneousPower);
+  }
+};
+
+SmartPlug.prototype.updateOn = function(on) {
+  this.properties.on = on;
+  if (on === null) {
+    return;
+  }
+
+  if (this.details) {
+    this.details.on.update();
+  }
+
+  if (on) {
+    this.showOn();
+    this.updatePower();
+  } else {
+    this.showOff();
+    this.smartPlugLabel.innerText = 'off';
+  }
+};
+
+/**
+ * Show updated power consumption value.
+ */
+SmartPlug.prototype.updatePower = function(power) {
+  this.properties.power = power;
+  if (!this.properties.on) {
+    this.smartPlugLabel.innerText = 'off';
+  } else if (power) {
+    this.smartPlugLabel.innerText = power + 'W';
+  } else {
+    this.smartPlugLabel.innerText = '...';
```

Comment:
I think it might be a bit clearer here to have 'on' because '...' is used by onOffSwitch to mean API failure